v3/lgr_hyper_ep.hpp

```diff
@@ -0,0 +1,648 @@
+// Allows for compile time branching in the material model.
+// The variables props_##ARG1 must be defined in lgr_hyper_ep_user.hpp.
+#define IF_MAT_PROPS_EQ(ARG1, ARG2) if constexpr(props_##ARG1 == ARG2)
+#define IF_MAT_PROPS_NEQ(ARG1, ARG2) if constexpr(props_##ARG1 != ARG2)
+#define IF_MAT_PROPS(ARG1) if constexpr(props_##ARG1)
+
+#else
+// Allows for run time branching in the material model.
+// The variables props.ARG1 are defined by reading the user input.
+#define IF_MAT_PROPS_EQ(ARG1, ARG2) if (props.ARG1 == ARG2)
+#define IF_MAT_PROPS_NEQ(ARG1, ARG2) if (props.ARG1 != ARG2)
+#define IF_MAT_PROPS(ARG1) if (props.ARG1)
+#endif
+
+#include <string>
+#include <limits>
+#include <algorithm>
+
+#include <hpc_symmetric3x3.hpp>
+
+namespace lgr {
+
+namespace hyper_ep {
+
+enum class ErrorCode {
+  NOT_SET,
+  SUCCESS,
+  LINEAR_ELASTIC_FAILURE,
+  HYPERELASTIC_FAILURE,
+  RADIAL_RETURN_FAILURE,
+  ELASTIC_DEFORMATION_UPDATE_FAILURE,
+  MODEL_EVAL_FAILURE
+};
+
+enum class Elastic { LINEAR_ELASTIC, NEO_HOOKEAN };
+
+enum class Hardening {
+  NONE,
+  LINEAR_ISOTROPIC,
+  POWER_LAW,
+  ZERILLI_ARMSTRONG,
+  JOHNSON_COOK
+};
+
+enum class RateDependence { NONE, ZERILLI_ARMSTRONG, JOHNSON_COOK };
+
+enum class Damage { NONE, JOHNSON_COOK };
+
+enum class StateFlag { NONE, TRIAL, ELASTIC, PLASTIC, REMAPPED };
+
+struct Properties {
+  // Elasticity
+  Elastic elastic;
+  double E;
+  double Nu;
+
+  // Plasticity
+  Hardening hardening;
+  RateDependence rate_dep;
+  double A;
+  double B;  // Hardening modulus
+  double n;  // exponent in hardening
+  double C1;
+  double C2;
+  double C3;
+  double C4;
+  double ep_dot_0;
+
+  // Damage parameters
+  Damage damage;
+  bool allow_no_tension;
+  bool allow_no_shear;
+  bool set_stress_to_zero;
+  double D1;
+  double D2;
+  double D3;
+  double D4;
+  double D5;
+  double D6;
+  double D7;
+  double D0;
+  double DC;
+  double eps_f_min;
+
+  Properties()
+      : elastic(Elastic::LINEAR_ELASTIC),
+        hardening(Hardening::NONE),
+        rate_dep(RateDependence::NONE),
+        damage(Damage::NONE),
+        allow_no_tension(true),
+        allow_no_shear(false),
+        set_stress_to_zero(false) {}
+};
+
+#ifdef LGR_COMPILE_TIME_MATERIAL_BRANCHES
+#include "lgr_hyper_ep_user.hpp"
+#endif
+
+inline char const* get_error_code_string(ErrorCode code) {
+  switch (code) {
+    case ErrorCode::NOT_SET:
+      return "NOT SET";
+    case ErrorCode::SUCCESS:
+      return "SUCCESS";
+    case ErrorCode::LINEAR_ELASTIC_FAILURE:
+      return "LINEAR ELASTIC FAILURE";
+    case ErrorCode::HYPERELASTIC_FAILURE:
+      return "HYPERELASTIC FAILURE";
+    case ErrorCode::RADIAL_RETURN_FAILURE:
+      return "RADIAL RETURN FAILURE";
+    case ErrorCode::ELASTIC_DEFORMATION_UPDATE_FAILURE:
+      return "ELASTIC DEFORMATION UPDATE FAILURE";
+    case ErrorCode::MODEL_EVAL_FAILURE:
+      return "MODEL EVAL FAILURE";
+  }
+  return "UNKNOWN";
+}
+
+template<class T>
```

Comment:
@ibaned this model is broken because of the function below - we need to port the `sqrt_spd` function from `Omega_h` to `lgr/v3`

Reply:
yep! we'll make that a priority